ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,325 @@
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    bool rootFound = false;
+    rootHandle = RootHandle_None;
+
+    RWLock_AcquireShared(s_entriesLock);
+    {
+        uintptr_t vnodeIndex;
+        if (TryFindVnodeIndex_Locked(vnode, vnodeHashIndex, /*out*/ vnodeIndex))
+        {
+            if (vnode == s_entries[vnodeIndex].vnode && vnodeVid == s_entries[vnodeIndex].vid)
+            {
+                rootFound = true;
+                rootHandle = s_entries[vnodeIndex].virtualizationRoot;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_entriesLock);
+    
+    return rootFound;
+}
+
+KEXT_STATIC void LookupVnodeRootAndUpdateCache(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    uint32_t vnodeVid,
+    bool forceRefreshEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    rootHandle = VirtualizationRoot_FindForVnode(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        vnode,
+        context);
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        if (!TryInsertOrUpdateEntry_ExclusiveLocked(
+                vnode,
+                vnodeHashIndex,
+                vnodeVid,
+                forceRefreshEntry,
+                rootHandle))
+        {
+            // TryInsertOrUpdateEntry_ExclusiveLocked can only fail if the cache is full
+            
+            perfTracer->IncrementCount(PrjFSPerfCounter_CacheFullCount, true /*ignoreSampling*/);
+        
+            InvalidateCache_ExclusiveLocked();
+            if (!TryInsertOrUpdateEntry_ExclusiveLocked(
+                        vnode,
+                        vnodeHashIndex,
+                        vnodeVid,
+                        true, // invalidateEntry
+                        rootHandle))
+            {
+                KextLog_FileError(
+                    vnode,
+                    "LookupVnodeRootAndUpdateCache: failed to insert vnode (%p:%u) after emptying cache",
+                    KextLog_Unslide(vnode), vnodeVid);
+            }
+        }
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC bool TryFindVnodeIndex_Locked(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    /* out parameters */
+    uintptr_t& vnodeIndex)
+{
+    // Walk from the starting index until we do one of the following:
+    //    -> Find the vnode
+    //    -> Find where the vnode should be inserted (i.e. NULLVP)
+    //    -> Have looped all the way back to where we started
+    vnodeIndex = vnodeHashIndex;
+    while (vnode != s_entries[vnodeIndex].vnode && NULLVP != s_entries[vnodeIndex].vnode)
+    {
+        vnodeIndex = (vnodeIndex + 1) % s_entriesCapacity;
```

Comment:
I'd check the emitted assembly, but you might want to change this to something like:

```
++vnodeIndex;
if (vnodeIndex >= s_entriesCapacity)
  vnodeIndex -= s_entriesCapacity;
```

or even 


```
++vnodeIndex;
if (vnodeIndex == s_entriesCapacity)
  vnodeIndex = 0;
```

Because we know that we'd only ever overflow in that way, and integer division by a variable denominator is very expensive. (Whereas the above should compile down to a conditional move, which is cheap.)

Reply:
I really like 

```
++vnodeIndex;
if (vnodeIndex == s_entriesCapacity)
  vnodeIndex = 0;
```

as an option since it's super clear, arguably clearer than what currently exists.